File: greedy/interval_cross/non_overlapping_intervals.go
package intervalcross

import "sort"

/**
本题区间排序使用的是区间的右边界，按照右边界升序排列，因为右边界越小，表示在右边留下的空间就越大
或者是按照区间的左边界排序，左边界越大，则表示在左边留下的空间就越大
*/
func eraseOverlapIntervals(intervals [][]int) int {
	// 根据区间的右边界升序排列
	sort.Slice(intervals, func(i, j int) bool {
		if intervals[i][1] == intervals[j][1] {
			return intervals[i][0] < intervals[j][0]
		}
		return intervals[i][1] < intervals[j][1]
	})

	var nonCross int = 1
	var edge int = intervals[0][1]
	for i := 1; i < len(intervals); i++ {
		if intervals[i][0] >= edge {
			nonCross++
			edge = intervals[i][1]
		}
	}

	return len(intervals) - nonCross
}

File: greedy/interval_cross/NonOverlappingIntervals.java
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2022-11-22 10:44:47
 * @LastEditTime: 2023-02-22 10:35:51
 */
package greedy.interval_cross;

import java.util.Arrays;

/**
 * leetcode 453 middle 无重叠区间
 * 
 * 给定一个区间的集合 intervals ，其中 intervals[i] = [starti, endi] 。
 * 返回 需要移除区间的最小数量，使剩余区间互不重叠 。
 * 
 *  [[1,2],[2,3],[3,4],[1,3]] 去掉[1,3]
 */
public class NonOverlappingIntervals {
    public int eraseOverlapIntervals(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> {
            if (a[0] == b[0]) {
                return a[1] - b[1];
            }
            return a[0] - b[0];
        });

        int[] bound = intervals[0];
        int ans = 0;
        for (int i = 1; i < intervals.length; i++) {
            if (intervals[i][0] < bound[1]) {
                ans++;
                bound[0] = intervals[i][0];
                bound[1] = Math.min(bound[1], intervals[i][1]);
            } else {
                bound = intervals[i];
            }
        }

        return ans;
    }
}
